website/blog/2023-02-21.ferretdb-compare-arrays.md

```diff
@@ -0,0 +1,530 @@
+7. ObjectId
+8. Boolean
+9. Date
+10. Timestamp
+11. Regular Expression
+
+The BSON comparison order is used for comparing different BSON types.
+A BSON array
+may use an element from an array to compare with other BSON type.
+We will discuss in detail in the following section.
+
+All number types are the same BSON comparison order, regardless of
+`double`, `int32`, `int64` or `decimal`.
+
+### Comparison Rules of Filter
+
+When comparing an array filter argument with array collections,
+each index is iterated in order.
+At each index, BSON comparison order is used then values are compared.
+If the values at a particular index are not equal, that result is used as the final comparison result.
+If there are no more elements in the array during the iteration,
+then that array is considered smaller.
+
+When using a filter argument comparing to an array,
+an element from that array would be selected.
+This is the case for both implicit operator and `$eq` operator.
+
+For operators `$lt` and `$lte` the smallest element from the array in the collection
+is used to compare against the filter argument.
+For operators `$gt` and `$gte` the largest element is used for comparison instead.
+Additionally, `$lt`, `$gt`, `$lte` and `$gte` uses BSON comparison order
+to only select elements of the same BSON comparison order as the filter argument.
+For example if the filter argument is string, only strings are selected.
+
+Filtering used in [`$lt`](#lt-operator-comparison) and [`$gt`](#gt-operator-comparison) operators
+applies the following rules to compare array and other BSON types.
+
+* `$lt` uses the smallest element of the same BSON type as non-array from the array.
+* `$gt` uses the largest element of the same BSON type as non-array from the array.
+* Empty array should not match other BSON type.
+
+### Example of Comparison Rules of Filter
+
+#### Example of Comparing Arrays
+
+To compare array filter argument and array in the collection arrays,
+we compare each element in the arrays until we discover a difference.
+For instance the following steps show `$lt` filter for comparing
+the filter argument `[3, 1, 'b']` with `[3, 'b', 'a']`.
+
+1. Compare `[3, 1, 'b']` and `[3, 'b', 'a']` using comparison order.
+   They have the same BSON type; both are arrays.
+2. Compare first index `3` and `3` using comparison order, both are numbers, and they are the same.
+3. Compare next index, `1` and `'b'` using comparison order,
+`1` is a number and less than string `'b'`.
+
+So `[3, 1, 'b']` is less than `[3, 'b', 'a']`.
+
+If an array has no more element to compare, the one with no more element is less.
+For the same reason empty array is less than arrays with elements.
+
+#### Example of Comparing Array and Other BSON Type
+
+For comparing non-array filter argument with array in the collection, `$lt` filter uses
+the smallest element from the array.
+The following steps show `$lt` filter for comparing filter argument `2`
+and `[3, 1]` in the collection:
```

Comment:
I would add a string or different value to this array, and present how filter will exclude non-matching types. So the filter would still be `{$lt:2}`, but collection would look like `[3,"foo",1]` or similar. The same for `$gt`

Reply:
Yea sure!